include/jwt.h

```diff
@@ -667,6 +670,202 @@ JWT_EXPORT void jwt_get_alloc(jwt_malloc_t *pmalloc, jwt_realloc_t *prealloc, jw
 
  /** @} */
 
+/**
+ * @defgroup jwt_vaildate JWT validation functions
+ * These functions allow you to define requirements for JWT validation.
+ *
+ * The most basic validation is that the JWT uses the expected algorithm.
+ *
+ * When replicating claims in header (usually for encrypted JWT), validation
+ * tests that they match claims in the body (iss, sub, aud).
+ *
+ * Time-based claims can also be validated (nbf, exp).
+ *
+ * Finally, validation can test that claims be present and have certain value.
+ *
+ * @{
+ */
+
+/**
+ * Validate a JWT object with a validation object.
+ *
+ * @param jwt Pointer to a JWT object.
+ * @param jwt_valid Pointer to a JWT validation object.
+ *
+ * @return 1 on valid, 0 on invalid, -1 on error
+ */
+JWT_EXPORT int jwt_validate(jwt_t *jwt, jwt_valid_t *jwt_valid);
+
+/**
+ * Allocate a new, JWT validation object.
+ *
+ * This is used to create a new object for a JWT validation. After you have
+ * finished with the object, use jwt_valid_free() to clean up the memory used by
+ * it.
+ *
+ * @param jwt_valid Pointer to a JWT validation object pointer. Will be allocated
+ *     on success.
+ * @return 0 on success, valid errno otherwise.
+ *
+ */
+JWT_EXPORT int jwt_valid_new(jwt_valid_t **jwt_valid, jwt_alg_t alg);
+
+/**
+ * Free a JWT validation object and any other resources it is using.
+ *
+ * After calling, the JWT validation object referenced will no longer be valid
+ * and its memory will be freed.
+ *
+ * @param jwt_valid Pointer to a JWT validation object previously created with
+ *     jwt_valid_new().
+ */
+JWT_EXPORT void jwt_valid_free(jwt_valid_t *jwt_valid);
+
+/**
+ * Add a new string grant requirement to this JWT validation object.
+ *
+ * @param jwt_valid Pointer to a JWT validation object.
+ * @param grant String containing the name of the grant to add.
+ * @param val String containing the value to be saved for grant. Can be
+ *     an empty string, but cannot be NULL.
+ * @return Returns 0 on success, valid errno otherwise.
+ *
+ * Note, this only allows for string based grants. If you wish to add
+ * integer grants, then use jwt_valid_add_required_grant_int(). If you wish to add more
+ * complex grants (e.g. an array), then use jwt_add_required_grants_json().
+ */
+JWT_EXPORT int jwt_valid_add_required_grant(jwt_valid_t *jwt_valid, const char *grant, const char *val);
```

Comment:
I think we can remove the `_required_` part of all of these functions. Since a validation object is meant to validate, it's probably safe to assume any grants added to it are required.